feat(taskfiles): Add cmake and remote utils; Update checksum data path to an array.

davidlion opened the pull request:
# Description
This PR adds two new sets of utility tasks, while also updating the checksum util tasks.

#### cmake utils
This PR adds two new tasks related to building a cmake project. The first runs build and configure, while the second performs the installation. Separating these tasks allows up to build a project once, but install it multiple times to different locations.

#### remote utils
The main utility task is `fetch-src` which will download and extract a remote tar ball. The `curl` task is separated out to allow directly downloading any files.

#### checksum utils
Previously the checksum utils only operated on a single data path, which is an issue for tasks that generate files across multiple directories. This PR just changes the arguments to be an array of paths rather than a single path.


# Validation performed
Used in ffi-go irv2 beta branch.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced checksum computation and validation tasks now support multiple input paths using wildcard patterns.
	- Introduced new tasks for CMake operations (`cmake-config-and-build`, `cmake-install`) and remote file handling (`curl`, `download-and-extract-tar`).
- **Bug Fixes**
	- Improved error handling to prevent task failures during checksum operations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

junhaoliao commented on taskfiles/utils.yml:
```diff
@@ -48,16 +50,19 @@ tasks:
+            ref: ".DATA_PATHS"
           EXCLUDE_PATHS:
-            ref: "default (list) .EXCLUDE_PATHS"
+            ref: ".EXCLUDE_PATHS"
```
I believe this default fallback is needed for Task `3.40.0`. Was the removal intentional?
```suggestion
            ref: "default (list) .EXCLUDE_PATHS"
```

davidlion replied:
No sorry. I think I might need to add this in other places too.

junhaoliao commented on taskfiles/utils.yml:
```diff
@@ -48,16 +50,19 @@ tasks:
+            ref: ".EXCLUDE_PATHS"
           OUTPUT_FILE: "{{.TMP_CHECKSUM_FILE}}"
       - defer: "rm -f '{{.TMP_CHECKSUM_FILE}}'"
       # Check that the directory exists and the checksum matches; otherwise delete the checksum file
```
```suggestion
      # Check that the path exists and the checksum matches; otherwise delete the checksum file
```

junhaoliao commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,159 @@ tasks:
+    internal: true
+    vars:
+      CHECKSUM_FILE: '{{default (printf "%s.md5" .BUILD_DIR) .CHECKSUM_FILE}}'
+      CMAKE_ARGS: "{{default nil .CMAKE_ARGS}}"
```
Would it be better to fallback to `""`?

davidlion replied:
Sure we can. I don't know if it makes a difference, but I don't think it can hurt either.

kirkrodrigues reviewed (changes requested):
Some quick high-level comments before a detailed review:

* The Rabbit has some suggestions that I think are valid.
* Can `fetch-src` replace CLP's [download-and-extract-tar](https://github.com/y-scope/clp/blob/53c4f52b3ae2ef0e990a8d66a7be87fd1e85cfbe/Taskfile.yml#L329)? If so, can we rename `fetch-src` to `download-and-extract-tar`?
* What docstring format are you basing things on?

davidlion commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,159 @@ tasks:
+          CHECKSUM_FILE: "{{.CHECKSUM_FILE}}"
+          DATA_PATHS: ["{{.OUTPUT_DIR}}"]
+    cmds:
+      - |-
+        rm -rf "{{.OUTPUT_DIR}}"
+        mkdir -p "{{.OUTPUT_DIR}}"
+        tar -x --strip-components="{{.STRIP}}" -C "{{.OUTPUT_DIR}}" -f "{{.TAR_FILE}}"
```
Task will already fail if any command fails.

davidlion commented:
> Some quick high-level comments before a detailed review:
> 
>     * The Rabbit has some suggestions that I think are valid.
> 
>     * Can `fetch-src` replace CLP's [download-and-extract-tar](https://github.com/y-scope/clp/blob/53c4f52b3ae2ef0e990a8d66a7be87fd1e85cfbe/Taskfile.yml#L329)? If so, can we rename `fetch-src` to `download-and-extract-tar`?
> 
>     * What docstring format are you basing things on?

New commits have added in Rabbit's suggestions, updated `fetch-src` to provide functionality of `download-and-extract-tar` (and renamed it), and updated docstrings based on disucssion.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -5,20 +5,19 @@ tasks:
+      vars: ["DATA_PATHS", "OUTPUT_FILE"]
     cmds:
       - >-
         tar cf -
```
Should we add `--wildcards` and `--no-anchored` like you do for the `download-and-extract-tar` task?

kirkrodrigues commented:
Review incoming.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -27,15 +26,18 @@ tasks:
   # @param {string} OUTPUT_FILE
-  # @param {[]string} [EXCLUDE_PATHS] A list of paths, relative to `DATA_DIR`, to exclude from the
-  # checksum.
+  # @param {string[]} [EXCLUDE_PATHS] List of paths, relative to any `DATA_PATHS`, to exclude from
```
Ditto suggestions from `compute-checksum`'s docstring.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -44,21 +46,24 @@ tasks:
           OUTPUT_FILE: "{{.TMP_CHECKSUM_FILE}}"
       - defer: "rm -f '{{.TMP_CHECKSUM_FILE}}'"
-      # Check that the directory exists and the checksum matches; otherwise delete the checksum file
+      # Check that the path exists and the checksum matches; otherwise delete the checksum file
```
```suggestion
      # Check that the paths exist and the checksum matches; otherwise delete the checksum file.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # CMAKE UTILS
+  # ===
+
+  # Runs cmake configure and build steps for the given source and build directories.
```
```suggestion
  # Runs CMake's configure and build steps for the given source and build directories.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+
+  # Runs cmake configure and build steps for the given source and build directories.
+  #
+  # @param {string} BUILD_DIR Cmake build directory to create.
```
```suggestion
  # @param {string} BUILD_DIR CMake build directory to create.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # @param {string} BUILD_DIR Cmake build directory to create.
+  # @param {string} SOURCE_DIR Project source directory containing the CMakeLists.txt file.
+  # @param {string={{.BUILD_DIR}}.md5} [CHECKSUM_FILE] Path to store the checksum of built files.
+  # @param {string=""} [CMAKE_ARGS] Any additional arguments to pass to cmake configure.
```
```suggestion
  # @param {string=""} [CMAKE_ARGS] Any additional arguments to pass to CMake's configure step.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # @param {string} SOURCE_DIR Project source directory containing the CMakeLists.txt file.
+  # @param {string={{.BUILD_DIR}}.md5} [CHECKSUM_FILE] Path to store the checksum of built files.
+  # @param {string=""} [CMAKE_ARGS] Any additional arguments to pass to cmake configure.
+  cmake-build:
```
Can we rename this to `cmake-config-and-build`?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+    requires:
+      vars: ["BUILD_DIR", "SOURCE_DIR"]
+    sources:
+      - "{{.SOURCE_DIR}}/**/*"
```
Note: Currently, this won't work for clp since the top-level CMakeLists.txt is in a directory with a lot of other files, but I guess the best solution is move all the CMakeLists and source files into a dedicated directory.

kirkrodrigues replied:
> What do you think about changing SOURCE_DIR to SOURCE_PATHS and default CMAKELISTS_PATH to SOURCE_PATHS[0]?

Hm, would that work with `sources`? Like is there a way to pass a list variable to `sources`?

davidlion replied:
Hmm good point... I'll look into it.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+      vars: ["BUILD_DIR", "SOURCE_DIR"]
+    sources:
+      - "{{.SOURCE_DIR}}/**/*"
+      - exclude: "{{.SOURCE_DIR}}/.cache/**/*"
```
What is this and what's generating it?

davidlion replied:
Sorry, I'll remove this. It is old code from when I was figuring out task. Iirc it is from my vim linting setup I think.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+    sources:
+      - "{{.SOURCE_DIR}}/**/*"
+      - exclude: "{{.SOURCE_DIR}}/.cache/**/*"
+      - exclude: "{{.SOURCE_DIR}}/compile_commands.json"
```
Isn't this generated in the build directory?

davidlion replied:
Sorry, I'll remove this. It is old code from when I was figuring out task.

You're right, but this line was aimed at a symlink related to my vim linting setup.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+      - task: "compute-checksum"
+        vars:
+          DATA_PATHS: ["{{.BUILD_DIR}}"]
+          OUTPUT_FILE: "{{.CHECKSUM_FILE}}"
+          EXCLUDE_PATHS: ["install_manifest.txt"]
```
Nit:
```suggestion
          EXCLUDE_PATHS: ["install_manifest.txt"]
          OUTPUT_FILE: "{{.CHECKSUM_FILE}}"
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+          OUTPUT_FILE: "{{.CHECKSUM_FILE}}"
+          EXCLUDE_PATHS: ["install_manifest.txt"]
+
+  # Runs cmake install step for the given build directory.
```
```suggestion
  # Runs the CMake install step for the given build directory.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+
+  # Runs cmake install step for the given build directory.
+  #
+  # @param {string} BUILD_DIR Cmake build directory.
```
```suggestion
  # @param {string} BUILD_DIR CMake build directory.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # Runs cmake install step for the given build directory.
+  #
+  # @param {string} BUILD_DIR Cmake build directory.
+  # @param {string} INSTALL_PREFIX Path prefix for installing the project.
```
```suggestion
  # @param {string} INSTALL_PREFIX Path prefix of where the project should be installed.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # @param {string} INSTALL_PREFIX Path prefix for installing the project.
+  # @param {string={{.INSTALL_PREFIX}}.md5} [CHECKSUM_FILE] Path to store the checksum of installed
+  # files.
+  # @param {string[]=[{{.INSTALL_PREFIX}}]} [DATA_PATHS] Paths to compute the the checksum for.
```
```suggestion
  # @param {string[]=[{{.INSTALL_PREFIX}}]} [DATA_PATHS] Paths to compute the checksum for.
```

davidlion replied:
I might be missing something, but this is what I've been going off of:

Without allowing the caller to specify the install prefix, we would always install to the system path by default. Installing to the system path could lead to conflicts or permission issues. The user could manually sets the `CMAKE_INSTALL_PREFIX` environment variable before running `task` but this would blindly apply to all `cmake-install` tasks.

For example: in ffi-go when we build the dependencies, I install them into the build directory (rather than the system) to ensure there are no conflicts and easy clean up.

ref: https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html

kirkrodrigues replied:
Oh I'm fine with `INSTALL_PREFIX`---I mean what do we need `DATA_PATHS` for?

davidlion replied:
Sorry too asleep when I responded lol.

We only want to compute the checksum for files we install, but a user might want to install to a file with other files. 
For example, say install prefix is `/usr/local` and we're installing library `foo`. We could end up with:
```
/usr/local/lib/bar
/usr/local/lib/foo
```

In a perfect world we could know what was installed from CMake (e.g. `lib/foo`), but that sounds hard :(.

Alternatively, I guess we could just always install and don't bother computing a checksum or using generates?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+    vars:
+      CHECKSUM_FILE: >-
+        {{default (printf "%s.md5" .INSTALL_PREFIX) .CHECKSUM_FILE}}
+      # Convert the install prefix to a single element array in the default case where DATA_PATHS
+      # has not been set.
+      _INSTALL_PREFIX_ARRAY: ["{{.INSTALL_PREFIX}}"]
+      DATA_PATHS:
+        ref: "default ._INSTALL_PREFIX_ARRAY .DATA_PATHS"
```
```suggestion
      DATA_PATHS:
        ref: "default (list .INSTALL_PREFIX) .DATA_PATHS"
```

I think that's equivalent?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # REMOTE UTILS
+  # ===
+
+  # Runs curl to download the provided URL.
```
```suggestion
  # Runs curl to download a file from the given URL.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  # Runs curl to download the provided URL.
+  #
+  # @param {string} URL
+  # @param {string} URL_SHA256 Content hash to verify downloaded file against.
```
How about `FILE_SHA256`?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  #
+  # @param {string} URL
+  # @param {string} URL_SHA256 Content hash to verify downloaded file against.
+  # @param {string={{(base .URL)}}} [OUTPUT_FILE] File path to store the download file.
```
```suggestion
  # @param {string={{(base .URL)}}} [OUTPUT_FILE] Path where the file should be stored.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+      OUTPUT_FILE: "{{default (base .URL) .OUTPUT_FILE}}"
+    requires:
+      vars: ["URL", "URL_SHA256"]
+    generates:
+      - "{{.OUTPUT_FILE}}"
```
```suggestion
    generates: ["{{.OUTPUT_FILE}}"]
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+          then
+            break
+          fi
+          echo "Attempt $attempt failed. Retrying..."
```
Won't this print too many times? Another way to write it is to move the loop breaking condition down here. (Fwiw, I feel like I've seen this code before and commented on the same issue. I guess it's from a StackOverflow post, lol.)

davidlion replied:
I'm just going to remove the log. The final log covers what happened in a failure anyway.

>Another way to write it is to move the loop breaking condition down here.
Fwiw, I'm pretty sure we'd need another condition to know whether we will retry again or not.

>(Fwiw, I feel like I've seen this code before and commented on the same issue. I guess it's from a StackOverflow post, lol.)
ya... it is from the rabbit

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+          exit 1
+        fi
+
+  # Runs curl to download the provided URL and tar to extract its contents.
```
```suggestion
  # Runs curl to download the tarball from the given URL and extracts its contents.
```

davidlion commented on taskfiles/utils.yml:
```diff
@@ -141,3 +146,194 @@ tasks:
+  #
+  # @param {string} OUTPUT_DIR Path to extract downloaded tar file contents to.
+  # @param {string} URL
+  # @param {string} URL_SHA256 Content hash to verify downloaded tar file against.
+  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
```
What do you think about `FILE_SHA256` to match the other task?

davidlion commented on taskfiles/utils.yml:
```diff
@@ -5,20 +5,19 @@ tasks:
   # @param {string} OUTPUT_FILE
-  # @param {[]string} [EXCLUDE_PATHS] A list of paths, relative to `DATA_DIR`, to exclude from the
-  # checksum.
+  # @param {string[]} [EXCLUDE_PATHS] List of paths, relative to any `DATA_PATHS`, to exclude from
+  # the checksum.
```
Want to change `DATA_PATHS` to `DATA_PATTERNS` to match?

kirkrodrigues replied:
Sure

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -5,37 +5,41 @@ tasks:
         {{- end}}
-        . 2> /dev/null
+        {{- range .DATA_PATTERNS}}
+        "{{.}}"
```
Will quoting a wildcard pattern work?

davidlion replied:
Good catch. It works for extract, but not create.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -44,21 +48,24 @@ tasks:
-        && diff -q '{{.TMP_CHECKSUM_FILE}}' '{{.CHECKSUM_FILE}}' 2> /dev/null
-        ) || rm -f '{{.CHECKSUM_FILE}}'
+        {{- range .DATA_PATTERNS}}
+        test -e "{{.}}" &&
```
(Ignoring the quotes) I probably should've anticipated this, but this won't work for patterns since if a wildcard path glob returns multiple files, `test -e` will fail because it only expects one argument.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +148,158 @@ tasks:
+        --build "{{.BUILD_DIR}}"
+        --parallel
+
+  # Unconditionally runs the CMake install step for the given build directory.
```
```suggestion
  # Runs the CMake install step for the given build directory.
```

Unless you think the unconditionally is important?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +148,158 @@ tasks:
+  # Runs curl to download a file from the given URL.
+  #
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify downloaded file against.
```
```suggestion
  # @param {string} FILE_SHA256 Content hash to verify the downloaded file against.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +148,158 @@ tasks:
+          exit 1
+        fi
+
+  # Runs curl to download the tarball file from the given URL and extracts its contents.
+  #
+  # @param {string} OUTPUT_DIR Directory in which to extract the tarball file.
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify downloaded tar file against.
+  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
+  # downloaded tar file.
+  # @param {string[]=[]} [EXCLUDE_PATTERNS] Path wildcard patterns that should not be extracted.
+  # @param {string[]=[]} [DATA_PATTERNS] Path wildcard patterns to extract.
+  # @param {int=1} [NUM_COMPONENTS_TO_STRIP] Number of leading path components to strip from the
+  # extracted files.
+  # @param {string={{.OUTPUT_DIR}}.tar.gz} [TAR_FILE] Path where the tarball file should be stored.
```
```suggestion
  # Runs curl to download the tar file from the given URL and extracts its contents.
  #
  # @param {string} OUTPUT_DIR Directory in which to extract the tar file.
  # @param {string} URL
  # @param {string} FILE_SHA256 Content hash to verify the downloaded tar file against.
  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
  # downloaded tar file.
  # @param {string[]=[]} [EXCLUDE_PATTERNS] Path wildcard patterns that should not be extracted.
  # @param {string[]=[]} [DATA_PATTERNS] Path wildcard patterns to extract.
  # @param {int=1} [NUM_COMPONENTS_TO_STRIP] Number of leading path components to strip from the
  # extracted files.
  # @param {string={{.OUTPUT_DIR}}.tar.gz} [TAR_FILE] Path where the tar file should be stored.
```

Upon re-reading, I realized `tar file` is better.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +148,158 @@ tasks:
+      TAR_FILE: >-
+        {{default (printf "%s.tar.gz" .OUTPUT_DIR) .TAR_FILE}}
+    requires:
+      vars: ["OUTPUT_DIR", "URL", "FILE_SHA256"]
```
Alphabetize

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +148,158 @@ tasks:
+    deps:
+      - task: "curl"
+        vars:
+          URL: "{{.URL}}"
+          FILE_SHA256: "{{.FILE_SHA256}}"
+          OUTPUT_FILE: "{{.TAR_FILE}}"
```
Alphabetize

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,77 @@
   # ===
 
-  # @param {string} DATA_DIR The directory to compute the checksum for.
+  # @param {string[]} DATA_PATTERNS Path wildcard patterns to compute the checksum for.
```
What do you think about `INCLUDE_PATTERNS` to match `EXCLUDE_PATTERNS`?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,77 @@
         --owner 0
         --sort name
-        {{- range .EXCLUDE_PATHS}}
+        --no-anchored
```
Can we add a comment about why this is necessary?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,77 @@
+        --no-anchored
+        --wildcards
+        {{- range .EXCLUDE_PATTERNS}}
         --exclude="{{.}}"
```
Can we add a comment about why the exclude pattern needs quotes but the include pattern doesn't?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,77 @@
-        ) || rm -f '{{.CHECKSUM_FILE}}'
+      # Check that all paths exist and the checksum matches; otherwise delete the checksum file.
+      - |-
+        ( {{- range .DATA_PATTERNS}}
```
```suggestion
        (
          {{- range .DATA_PATTERNS}}
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +154,158 @@ tasks:
+      - >-
+        cmake
+        --build "{{.BUILD_DIR}}"
+        --parallel
```
Can we add an optional arg to control the number of jobs (useful for memory constrained environments)?

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +154,158 @@ tasks:
+        | awk '{print $2}')
+    cmds:
+      - |-
+        mkdir -p "{{dir .OUTPUT_FILE}}"
```
Not mandatory, but to follow our conventions, we should use `--parents`

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +154,158 @@ tasks:
+  # @param {string} OUTPUT_DIR Directory in which to extract the tar file.
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify the downloaded tar file against.
+  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
```
```suggestion
  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of the
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +154,158 @@ tasks:
+          DATA_PATTERNS: ["{{.OUTPUT_DIR}}"]
+    cmds:
+      - |-
+        rm -rf "{{.OUTPUT_DIR}}"
+        mkdir -p "{{.OUTPUT_DIR}}"
```
Not mandatory, but could use long form flags.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +154,158 @@ tasks:
+        {{default (printf "%s.tar.gz" .OUTPUT_DIR) .TAR_FILE}}
+    requires:
+      vars: ["FILE_SHA256", "OUTPUT_DIR", "URL"]
+    sources: ["{{.TASKFILE}}"]
+    generates: ["{{.CHECKSUM_FILE}}", "{{.TAR_FILE}}"]
```
I need to double-check if these will cause issues.

kirkrodrigues replied:
* I guess you want `.TASKFILE` to be a source so that if the way of extracting the file changes, it will be extracted again, right?
* I think `.TARFILE` should be a source (rather than a generated file) since it's the generated file of `curl`.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,87 @@
-        test -d "{{.DATA_DIR}}"
-        && diff -q '{{.TMP_CHECKSUM_FILE}}' '{{.CHECKSUM_FILE}}' 2> /dev/null
-        ) || rm -f '{{.CHECKSUM_FILE}}'
+          {{- range .DATA_PATTERNS}}
```
```suggestion
          {{- range .INCLUDE_PATTERNS}}
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,87 @@
     ignore_error: true
 
-  # @param {string} DATA_DIR The directory to validate the checksum for.
+  # @param {string[]} INCLUDE_PATTERNS Path wildcard patterns to validate the checksum for.
   # @param {string} OUTPUT_FILE
```
```suggestion
  # @param {string} CHECKSUM_FILE
  # @param {string[]} INCLUDE_PATTERNS Path wildcard patterns to validate the checksum for.
```

Note the rename from OUTPUT_FILE to CHECKSUM_FILE.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +164,164 @@ tasks:
+    internal: true
+    label: "{{.TASK}}-{{.OUTPUT_DIR}}"
+    vars:
+      CHECKSUM_FILE: >-
+        {{default (printf "%s.md5" .OUTPUT_DIR) .CHECKSUM_FILE}}
+      INCLUDE_PATTERNS:
+        ref: "default (list) .INCLUDE_PATTERNS"
+      EXCLUDE_PATTERNS:
+        ref: "default (list) .EXCLUDE_PATTERNS"
+      NUM_COMPONENTS_TO_STRIP: "{{default 1 .NUM_COMPONENTS_TO_STRIP}}"
+      TAR_FILE: >-
+        {{default (printf "%s.tar.gz" .OUTPUT_DIR) .TAR_FILE}}
```
```suggestion
      # Paths
      CHECKSUM_FILE: >-
        {{default (printf "%s.md5" .OUTPUT_DIR) .CHECKSUM_FILE}}
      TAR_FILE: >-
        {{default (printf "%s.tar.gz" .OUTPUT_DIR) .TAR_FILE}}
      
      # Path patterns
      EXCLUDE_PATTERNS:
        ref: "default (list) .EXCLUDE_PATTERNS"
      INCLUDE_PATTERNS:
        ref: "default (list) .INCLUDE_PATTERNS"
      
      # Misc
      NUM_COMPONENTS_TO_STRIP: "{{default 1 .NUM_COMPONENTS_TO_STRIP}}"
```

Our guidelines ask that [map entries be ordered alphabetically](https://docs.yscope.com/dev-guide/contrib-guides-taskfiles.html#ordering-of-list-map-entries), within each group, so I think we can break this into groups.

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,87 @@
-      vars: ["DATA_DIR", "OUTPUT_FILE"]
+      vars: ["INCLUDE_PATTERNS", "OUTPUT_FILE"]
     cmds:
+      # We explicitly set --no-anchored and --wildcards to ensure the behaviour for both exclusion
+      # and inclusion matches, as they are only the default for exclusion.
+      # We can quote the exclude patterns for tar because the patterns themselves are passed to tar
+      # as arguments and evaluated by tar during execution. However, for tar --create the input
+      # patterns cannot be quoted as they are evaluated by the shell and the results are passed to
+      # tar as arguments. If the input patterns are passed to tar with quotes the pattern will not
+      # be evaluated and will be treated literally.
```
```suggestion
      # We explicitly set `--no-anchored` and `--wildcards` to make the inclusion behaviour match
      # the default exclusion behaviour.
      #
      # We can quote the exclude patterns for `tar` because the patterns themselves are passed to
      # `tar` as arguments and evaluated by `tar` during execution. However, for `tar --create`, the
      # input patterns cannot be quoted since they're evaluated by the shell and the results are
      # passed to `tar` as arguments. If the input patterns are passed to `tar` with quotes, the
      # pattern won't be evaluated and will instead be treated literally.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -1,64 +1,87 @@
   # @param {string} OUTPUT_FILE
-  # @param {[]string} [EXCLUDE_PATHS] A list of paths, relative to `DATA_DIR`, to exclude from the
-  # checksum.
+  # @param {string[]} [EXCLUDE_PATTERNS] Path wildcard patterns, relative to any `INCLUDE_PATTERNS`, to
+  # exclude from the checksum.
```
```suggestion
  # @param {string[]} [EXCLUDE_PATTERNS] Path wildcard patterns, relative to any `INCLUDE_PATTERNS`,
  # to exclude from the checksum.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +164,164 @@ tasks:
+  # @param {string} BUILD_DIR CMake build directory to create.
+  # @param {string} SOURCE_DIR Project source directory containing the CMakeLists.txt file.
+  # @param {int} [JOBS] The maximum number of concurrent processes to use when building. If
+  # omitted the native build tool's default number is used. See man cmake.
```
```suggestion
  # omitted, the native build tool's default number is used. See `man cmake`.
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +164,164 @@ tasks:
+    internal: true
+    label: "{{.TASK}}-{{.SOURCE_DIR}}-{{.BUILD_DIR}}"
+    vars:
+      JOBS: >-
+        {{default "" .JOBS}}
+      CONF_ARGS: >-
+        {{default "" .CONF_ARGS}}
```
```suggestion
      CONF_ARGS: >-
        {{default "" .CONF_ARGS}}
      JOBS: >-
        {{default "" .JOBS}}
```

kirkrodrigues commented on taskfiles/utils.yml:
```diff
@@ -141,3 +164,164 @@ tasks:
+      - |-
+        rm --force --recursive "{{.OUTPUT_DIR}}"
+        mkdir --parents  "{{.OUTPUT_DIR}}"
+      # Unlike tar --create, in --extract both the exclude and include patterns can be quoted as
+      # they are both evaluted by tar.
```
```suggestion
      # Unlike for `tar --create`, in `--extract` both the exclude and include patterns can be
      # quoted since they're both evaluated by `tar`.
```

kirkrodrigues reviewed (changes requested):
Mainly style and one error.

For the PR title, how about:

```
feat(taskfiles): Add cmake and remote utils; Update checksum data path to an array.
```
